x-pack/plugins/cloud_security_posture/public/pages/findings/findings_container.tsx

```diff
@@ -5,66 +5,107 @@
-import type { CSPFinding, FetchState } from './types';
-import type { CspPluginSetup } from '../../types';
+import { TEST_SUBJECTS } from './constants';
+import {
+  extractErrorMessage,
+  useSourceQueryParam,
+  useEsClientMutation,
+  isNonNullable,
+} from './utils';
+import type { CSPFinding, FindingsFetchState } from './types';
+import type { DataView, IKibanaSearchResponse } from '../../../../../../src/plugins/data/common';
+import type { SearchBarProps } from '../../../../../../src/plugins/data/public';
+import { useKibana } from '../../../../../../src/plugins/kibana_react/public';
+
+type FindingsEsSearchMutation = UseMutationResult<
+  IKibanaSearchResponse<SearchResponse<CSPFinding>>,
+  unknown,
+  void
+>;
+
+export type URLState = Parameters<NonNullable<SearchBarProps['onQuerySubmit']>>[0] & {
+  filters: Filter[];
+};
+
+const getDefaultQuery = (): Required<URLState> => ({
+  query: { language: 'kuery', query: '' },
+  filters: [],
+  dateRange: {
+    from: 'now-15m',
+    to: 'now',
+  },
+});
+
+// with https://github.com/microsoft/TypeScript/pull/46266
+// destructuring would make this more concise and won't leak props
+export const getFetchState = <T extends FindingsEsSearchMutation>(v: T): FindingsFetchState => {
+  switch (v.status) {
+    case 'error':
+      return { ...v, error: extractErrorMessage(v.error) };
+    case 'success':
+      return {
+        ...v,
+        data: v.data?.rawResponse?.hits?.hits?.map((h) => h._source).filter(isNonNullable),
+      };
+    default:
+      return v;
+  }
+};
 
 /**
  * This component syncs the FindingsTable with FindingsSearchBar
  */
-export const FindingsTableContainer = () => {
-  const { kubebeatDataView } = useKubebeatDataView();
-  const [state, set] = useState<FetchState<CSPFinding[]>>({
-    loading: false,
-    error: false,
-    data: [],
+export const FindingsTableContainer = ({ dataView }: { dataView: DataView }) => {
+  const { notifications } = useKibana().services;
+  const [selectedFindingsItem, selectItem] = useState<CSPFinding | undefined>();
+  const { source: searchState, setSource: setSearchSource } = useSourceQueryParam(getDefaultQuery);
+  const mutation = useEsClientMutation<CSPFinding>({
+    ...searchState,
+    dataView,
   });
+  const fetchState = getFetchState(mutation);
 
-  const onError = useCallback((v) => set({ error: v, loading: false, data: undefined }), []);
-  const onSuccess = useCallback((v) => set({ error: false, loading: false, data: v }), []);
-  const onLoading = useCallback(() => set({ error: false, loading: true, data: undefined }), []);
-
-  if (!kubebeatDataView) return null;
+  // This sends a new search request to ES
+  // it's called whenever we have a new searchState from the URL
+  useEffect(() => {
+    mutation.mutate(void 0, {
+      onError: (e) => {
+        notifications?.toasts.addError(e instanceof Error ? e : new Error(), {
+          title: 'Search failed',
+        });
```

Comment:
cool stuff, kinda thing i would like to have in a util function so we can easily reuse without remembering the exact spell 🧙‍♀️

```suggestion
// other realm
export const getErrorOrCustom = (e, custom) => e instanceof Error ? e : new Error(), { title: custom }

// findings_container
onError: (e) => notifications?.toasts.addError(getErrorOrCustom(e, 'Search Failed'));
```

Reply:
we'll have to add it for sure if there isn't already one

i'm not sure this is the right way to go tho -
1 - it's only semi-convenient if we add a utility, so this hints something may be done wrong here 
2 - creating Errors from thin air isn't really constructive (haha! ), as it won't have an informative `message` and its callstack won't point to the actual origin of the error.

Reply:
yeah correct. maybe the whole `notification?.toasts.addError` should be a `useKibanaErrorToast`?

Reply:
yeah, that'd be more fitting. but i think there has to be one already, no ?